### ISSUE TITLE
EVP_PKEY_new_raw_private_key fails for a zero length key

### ISSUE THREAD

**Issue by gotthardp** [openssl/openssl#15632]
```
EVP_PKEY *pkey = EVP_PKEY_new_raw_private_key(EVP_PKEY_HMAC, NULL, "", 0);
```
returns a valid pointer in OpenSSL 1.1.1,

but returns `NULL` in OpenSSL 3.0 with the following error stack
```
80BB21C91F7F0000:error:1C8C0100:Provider routines:mac_key_fromdata:malloc failure::0:
80BB21C91F7F0000:error:030000B4:digital envelope routines:new_raw_key_int:key setup failed::0:
```
This is probably because `OPENSSL_secure_malloc(0)` returned `NULL`, which does **not** indicate an error.

**Comment by slontis** [comment]
I have to ask.. Why would you do that? :)

**Comment by gotthardp** [comment]
This is how TPM implements empty passwords.

The TCG specification for TPM says e.g. *"The EmptyAuth value may be provided in either an HMAC session using a zero-length authValue in the HMAC calculation or as a zero length password."*
The current implementation in https://github.com/tpm2-software/tpm2-tss uses a zero-length key HMAC calculation to do that.

The RFC 2104 says the *"key for HMAC can be of any length"*, so this feels to be a valid behaviour.

**Comment by slontis** [comment]
Probably should work how it did before..
EVP_MAC would be a good idea to swap to at some point.

**Comment by gotthardp** [comment]
Yeah. The disadvantage of EVP_MAC is that it works with 3.0 only.

**Comment by t8m** [comment]
Fix in #15643
